pkg/schedule/schedule.go

```diff
@@ -149,17 +156,29 @@ func (f *fifo) run() {
 				f.mu.Unlock()
 				// clean up pending jobs
 				for _, todo := range pendings {
-					todo(f.ctx)
+					f.executeJob(todo, true)
 				}
 				return
 			}
 		} else {
-			todo(f.ctx)
+			f.executeJob(todo, false)
+		}
+	}
+}
+
+func (f *fifo) executeJob(todo Job, ctxDone bool) {
+	defer func() {
+		if !ctxDone {
 			f.finishCond.L.Lock()
 			f.finished++
 			f.pendings = f.pendings[1:]
 			f.finishCond.Broadcast()
 			f.finishCond.L.Unlock()
 		}
-	}
+		if err := recover(); err != nil {
```

Comment:
I have a doubts about the `catching exceptions and swallowing errors` behavior.
If one of the periodic book-keeping processes within DB server fails unexpectedly, it might be caused by assertion that is supposed to detect data inconsistency or other serious problem (e.g. during compaction).   The earlier we detect it and avoid propagation (by e.g. reading checkpoint state from the other replica), the lowest chances we will make this 'broken' state authoritative and the easier to debug the problem. 

How about expanding the logging (as you did), flushing the logs, but applying this pattern: 

https://stackoverflow.com/questions/34842322/can-i-recover-from-panic-handle-the-error-then-panic-again-and-keep-the-origin

instead of the `recover` ?

We should look at this error 1 by 1 and decide whether they are recoverable or not.

Reply:
Got it. About the `recover`, I write a demo.
![image](https://user-images.githubusercontent.com/21985684/173609449-6e7b9c9c-a3e1-4432-9495-264af0d83b94.png)

![image](https://user-images.githubusercontent.com/21985684/173609317-ebc27ef9-4c50-4502-8095-90931c2e259a.png)

So I think the recover method can be used. So I plan to add a `HandlePanic` func in the `Job` struct. When the `HandlePanic` isn't nil, it will be called when the panic happens.

Reply:
Thank you. This seems good overall. 

Minor doubt I have... is what's the benefit of having separate HandlePanic(...) if Job author could actually already handle it within the Job.Do itself.

Reply:
Thanks your reading. It seems that the `HandlePanic ` method is a little redundant.